Add Resolver#isSystemError

oguzhanunlu opened the pull request:
Add Resolver#isSystemError to allow users to determine if a resolution error is caused by Iglu server unavailability.

Returns true when:
- At least one custom Iglu Server is *unhealthy*
- All Iglu Central mirrors are *unhealthy*

`*unhealthy*`: errors of a LookupHistory can be either
- full of RepoFailure: easy to consider this registry as unhealthy
- a mix of RepoFailure along with NotFound and/or ClientFailure: I considered the registry unavailable, since errors is a Set[RegistryError] which is unordered
- no RepoFailure at all: can't consider registry as unhealthy

Returns false when:
- At least one Iglu Central mirror is healthy
- All custom Iglu Servers are healthy

ref: https://snplow.atlassian.net/browse/PDP-2318

oguzhanunlu committed:
Add Resolver#isUnrecoverable

Add Resolver#isUnrecoverable to allow users to determine if a resolution error is caused by Iglu server unavailability.

Returns true when:
- At least one custom Iglu Server is *unhealthy*
- All Iglu Central mirrors are *unhealthy*

`*unhealthy*`: errors of a LookupHistory can be either
- full of RepoFailure: easy to consider this registry as unhealthy
- a mix of RepoFailure along with NotFound and/or ClientFailure: I considered the registry unavailable, since errors is a Set[RegistryError] which is unordered
- no RepoFailure at all: can't consider registry as unhealthy

Returns false when:
- At least one Iglu Central mirror is healthy
- All custom Iglu Servers are healthy

istreeter commented on modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala:
```diff
@@ -205,6 +205,33 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
+    }
+
+    def hasRepoFailure(history: LookupHistory): Boolean =
+      history.errors.exists(_.isInstanceOf[RegistryError.RepoFailure])
```
There are three types of `RegistryError`.  Out of the three types, only `NotFound` is expected, i.e. a tracking error not a pipeline error.  Both of the other two (`RepoFailure` and `ClientError`) indicate some kind of issue with the pipeline, which should never happen.  How about we crash on `ClientError` too?  I cannot think of a reason not to.

Even the scaladoc for `ClientError` says [it is fatal](https://github.com/snowplow/iglu-scala-client/blob/4.1.0/modules/data/src/main/scala/com.snowplowanalytics.iglu/client/resolver/registries/RegistryError.scala#L31).

oguzhanunlu replied:
yeah that makes sense, I'll push a new commit for that

oguzhanunlu committed:
Treat ClientFailure as fatal in isUnrecoverable

Address code review feedback:
- ClientFailure is now treated as fatal/unrecoverable alongside RepoFailure
- Only NotFound errors are considered recoverable
- Renamed internal helper from hasRepoFailure to hasFatalError for clarity
- Expanded test coverage with clear naming conventions (f1-f9 for false cases, t1-t17 for true cases)

oguzhanunlu committed:
fmt

istreeter commented on modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala:
```diff
@@ -205,6 +205,35 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
+   *
+   * If a registry has a fatal error alongside other errors, assume it is unrecoverable.
+   */
+  def isUnrecoverable(error: ResolutionError): Boolean = {
```
I have doubts about the name `isUnrecoverable`.  Bearing in mind this becomes part of the public API of this library, so name is quite important.

If `isUnrecoverable` returns `false` then it implies this error is "recoverable".  But what does "recoverable" mean in this context?  It means we're going to send this event to the bad stream... but I wouldn't say that is an example of "being recoverable".

You know I'm not good at naming things :) so I don't have good suggestions to change the name.

Maybe `isSystemError` or `isPlatformError`... meaning the system/platform/deployment/configuration is the problem, and there is no problem with the data.

oguzhanunlu replied:
I think `isSystemError` sounds good, thanks for the suggestion!

istreeter commented on modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala:
```diff
@@ -205,6 +205,35 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
+    def hasFatalError(history: LookupHistory): Boolean =
+      history.errors.exists(e =>
+        e.isInstanceOf[RegistryError.RepoFailure] || e.isInstanceOf[RegistryError.ClientFailure]
+      )
```
This is fine.  My personal preference is to do something like:

```scala
history.errors.exists {
  case _: RegistryError.RepoFailure => true
  case _: RegistryError.ClientFailure => true
  case _: RegistryError.NotFound => false
}
```

...because that is more explicit to someone reading the code about how we categorize each of the error types.

But I don't want to nitpick, and other engineers will probably disagree with my preferences!  So how you did it is fine.

oguzhanunlu replied:
> because that is more explicit to someone reading the code about how we categorize each of the error types

exactly, I like it and will modify it. Thanks for the suggestion 👍

oguzhanunlu committed:
Rename isUnrecoverable to isSystemError

- Better naming: indicates system/platform issue vs data issue
- Use pattern matching for explicit error type categorization

istreeter commented on modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala:
```diff
@@ -205,6 +205,37 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
+   *
+   * Returns false otherwise (e.g., NotFound errors indicate missing schema, not system failure).
+   *
+   * If a registry has a fatal error alongside other errors, assume it is a system error.
```
I just noticed this comment.  Please can you help me understand exactly what happens if a registry has fatal errors AND non-fatal errors.

Imagine a pipeline has some mis-configured tracking, so it generates a lot of `NotFound` schema violations.  In this scenario the system is healthy, even though the tracking is wrong.  This is probably a very common scenario.

Now imagine the Iglu Server goes down for this pipeline and starts returning 5xx for all requests.  I think this an example where the `LookupHistory` will contain a mixture of fatal errors (`RepoFailure`) AND what you call "other" errors (`NotFound`).

When you think about this scenario... we don't want Enrich to crash in this scenario.  We have successfully cached the fact that this schema is not found in all repos.  So I think it's better to generate a schema violation with `NotFound` instead of crashing Enrich.

So.... can you describe what happens in that scenario I described?  Would Enrich crash or stay alive in that scenario?  And if you need to update the implementation and tests, then please do so.

P.S. Sorry I keep noticing new things to complain about!!

oguzhanunlu replied:
I'm glad that you question the approach early rather than late!

In your scenario, enrich would crash. Why? LookupHistory#errors is of type `Set[RegistryError]` which is unordered:

{NotFound, RepoFailure} could mean:
  - NotFound first → then server went down (RepoFailure) = server is currently down
  - RepoFailure first → then server recovered and said NotFound = server is healthy now

Since we can't distinguish, I went down the current route so that events wouldn't go to bad stream.

> So I think it's better to generate a schema violation with NotFound instead of crashing Enrich.

Yeah I agree. I put too much priority on the idea of `events shouldn't go to bad`. I'll change accordingly. Thanks for the heads up!

oguzhanunlu committed:
NotFound mixed with fatal errors should not be a system error

If a registry has NotFound alongside RepoFailure/ClientFailure, we now consider this NOT a system error. The presence of NotFound confirms the schema genuinely doesn't exist, so crashing won't help.

This prevents unnecessary crashes in the common scenario where tracking is misconfigured (NotFound) and then the Iglu Server goes down (RepoFailure).

oguzhanunlu committed:
fmt

istreeter reviewed (approved):
I think this is fine.

We should get the opinion of @benjben before releasing Enrich, because he previously thought about Iglu errors when working on a similar feature.  He might spot something I missed.

It's fine to merge this for now.  If @benjben has any suggestions for improvement then we can address it in a follow-up PR.

benjben commented:
I think that the algorithm looks great !

If for a repo we have first a failure and then a `NotFound` it works well as `NotFound` takes precedence over any other kind of failure.
